### ISSUE TITLE
Support for parsing of nodes with "mixed" content, i.e. containing text and element nodes

### ISSUE THREAD

**Issue by marmutro** [ymtszw/elm-xml-decode#9]
Hi Yu
I have to decode: `<Text lang="en">Press <uicontrol>Next</uicontrol> to save the values.</Text>`
The string decoder fails with "The node is not a simple text node." so I need a "mixedContent" decoder:
My proposal:

```
mixedContent : Decoder String
mixedContent =
    Decoder mixedContentImpl

mixedContentImpl : Node -> Result Error String
mixedContentImpl node =
    let
        nodesToString : List XmlParser.Node -> String
        nodesToString nodes =
            nodes
                |> List.map nodeToString
                |> String.join ""

        nodeToString : XmlParser.Node -> String
        nodeToString aNode =
            case node of
                XmlParser.Text str ->
                    str

                XmlParser.Element _ _ _ ->
                    XmlParser.format { docType = Nothing, processingInstructions = [], root = aNode }
    in
    case node of
        Text str ->
            Ok str

        Element _ _ nodes ->
            Ok (nodesToString nodes)

```
This decodes the above example into `"Press <uicontrol>Next</uicontrol> to save the values.".`
Would it be possible to add such a decoder?

Regards
Martin

**Comment by ymtszw** [comment]
Hi Martin! I'm pleased to see you've deeply utilized `elm-xml-decode`!

At a glance, it is unclear to me what kind of tags/elements are considered "special" in such cases and should be allowed. Is it:

- all tags appearing as direct children of your targeted `Node`? (*1)
- only select set of tag(s) appearing as direct children of your targeted `Node`? (*2)
  - In this case, unsupported tags, if any, should be treated as error, or their text contents should be extracted?
- or, one of the above with any-depth descendants of your targeted `Node`? (*3)

And also,

- Should we have "text-only" variants of such decoder? (*4)
  - I know it is slight off-topic, but i.e. extracting just texts from a `Node` and generating `"Press Next to save the values.".` could be useful?

Just a quick thought:
- (*1) may find its place as library API. Naming needs to be improved through.
- (*2) is also possible but feels a bit too convoluted to me. Although such features can be found in markdown parsers for instance, like allowing only a subset of HTML tags.
- (*3) is possible if (*2) is, but again, feels convoluted and seems somewhat lessens the power of decoders, because tags rendered in `String` are no longer decoded until another decoding process begins. Of course, switching in-and-out between decode-able `Node` and `String` is always possible, but exact situations can widely vary between users. (Creating their own purpose-built decoder could be faster.)
- (*4) is actually very reasonable to have, and I've even seen it before as XPath feature.

**Comment by marmutro** [comment]
Hi Yu
I played a bit with my suggestion in a local copy and now I think it would be wiser to have a decoder that just returns the parsed node (i.e XmlParser.Node) without converting it to text. Because later in the "rendering" code the text has to be parsed again for those uicontrol tags.
BTW your decoders allowed me to easily switch from JSON to Xml (Unfortunately I have to because the server prefers sending Mqtt-Topics in Xml).
I know that just exposing Decoder(..) is not a wise option.

**Comment by ymtszw** [comment]
> I played a bit with my suggestion in a local copy and now I think it would be wiser to have a decoder that just returns the parsed node (i.e XmlParser.Node) without converting it to text. Because later in the "rendering" code the text has to be parsed again for those uicontrol tags.

I see. In that case, decoder that "yielding `Node` without converting to other type" would be your need, right? I.e. [`Json.Decode.value`](https://package.elm-lang.org/packages/elm/json/latest/Json-Decode#value) for XML?

Signature would be `node : Decode XmlParser.Node`, and usage would be for instance:

```elm
textNodeExtractor : Decoder XmlParser.Node
textNodeExtractor =
    path [ "path", "to", "Text" ] (single node)

run textNodeExtractor  "...<XML doc>..."
--> XmlParser.Element "Text" [ { name = "lang", value = "en" } ] [ ...<children> ... ]
```

This is definitely a thing!

**Comment by marmutro** [comment]
For my use case (*1) would be sufficient.

**Comment by marmutro** [comment]
That would be perfect !
And a very simple implementation:
```
node : Decoder Node
node =
    Decoder Ok

```

**Comment by ymtszw** [comment]
Cool!

Though I want a test case or two (and function doc) in conjunction with `path` function, to demonstorate which `Node`(s) are returned when combined with either `list` or `single`.
In essence:

```elm
node = Decoder Ok

run (path ["path", "to", "target"] (single node)) """
<root>
  <path>
    <to>
      <target>
        <child></child>
        <child></child>
      </target>
      <target></target>
    </to>
  </path>
</root>
"""
--> I believe this returns `Element "target" [] [ Element "child" [] [], Element "child" [] [] ]`,
--> NOT `Element "child" [] []`.
--> But not certain 'cause I'm away from the code for a good long time :P
```

Please confirm me either way! Also a PR is welcome. Even if not be, I can tackle this on the weekend :)

**Comment by marmutro** [comment]
I just tried it in elm repl: "single" returns an Err "Multiple nodes found" because there are 2 "target" nodes
```
> run (path ["path", "to", "target"] (single node)) """<root><path><to><target><child></child><child></child></target><target></target></to></path></root>"""
Err ("Path: /path/to/target\nNode: <to><target><child></child><child></child></target><target></target></to>\nMultiple nodes found.")
    : Result String XmlParser.Node
```
But with "list" it returns Ok:
```
> run (path ["path", "to", "target"] (list node)) """<root><path><to><target><child></child><child></child></target><target></target></to></path></root>"""
Ok [Element "target" [] [Element "child" [] [],Element "child" [] []],Element "target" [] []]

```

**Comment by ymtszw** [comment]
> I just tried it in elm repl: "single" returns an Err "Multiple nodes found" because there are 2 "target" nodes
> 
> ```
> > run (path ["path", "to", "target"] (single node)) """<root><path><to><target><child></child><child></child></target><target></target></to></path></root>"""
> Err ("Path: /path/to/target\nNode: <to><target><child></child><child></child></target><target></target></to>\nMultiple nodes found.")
>     : Result String XmlParser.Node
> ```
> 
> But with "list" it returns Ok:
> 
> ```
> > run (path ["path", "to", "target"] (list node)) """<root><path><to><target><child></child><child></child></target><target></target></to></path></root>"""
> Ok [Element "target" [] [Element "child" [] [],Element "child" [] []],Element "target" [] []]
> ```

Oh I should've known lol. Thanks for clarifying that!

The behavior looks perfect. Please go ahead and file a PR 👍

**Comment by ymtszw** [comment]
Via #10 , releases soon!

**Comment by ymtszw** [comment]
Released as [3.2.0](https://package.elm-lang.org/packages/ymtszw/elm-xml-decode/3.2.0/)

**Comment by marmutro** [comment]
Maybe I mistyped the link to XmlParser. There is something odd in the doc of "node". The link to XmlParser is missing.

**Comment by ymtszw** [comment]
Fixed and pushed as 3.2.1
